### DISCUSSION TITLE
Unsafe Rubocop autofixes?

### DISCUSSION THREAD

**Discussion by stdedos** [rubocop/rubocop#13294] [Q&A] [answered]
_Title was: Is it possible that `Style/BlockDelimiters` "can be unsafe"?_

I did a massive `rubocop -a` (`rubocop (1.66.1)`) across a repository, and one of the changes is:

```diff
     g.field :field_name, N_('Field Name'), type: :string, editable: false,
-            options: ModuleName::OptionClass.options,
-            help: N_('Description of the field. Possible options are: %s') %
-              ModuleName::OptionClass.values.map { |t| t.get_label }.join(', ')
+                                                             options: ModuleName::OptionClass.options,
+                                                             help: N_('Description of the field. Possible options are: %s') %
+                                                                   ModuleName::OptionClass.values.map do |t|
+                                                                     t.get_label
+                                                                   end.join(', ')
```

(Hopefully this is _indeed_ `Style/BlockDelimiters` 😅)

"Building" leads to:
```
org.jruby.exceptions.NoMethodError: (NoMethodError) undefined method `join' for #<Field:0x3316aef6>
Did you mean?  JSON
```

Docs https://docs.rubocop.org/rubocop/cops_style.html#styleblockdelimiters say that the autofix should be safe.

Ruby so far for me is just a jungle I found myself in - please be kind to me if I am obviously missing something 🙏 
(PS: I _did_ read the doc page. But `--autofix` **MUST** be safe, at all times. I haven't been bitten so far e.g. by `ruff` or `black` autofix)

**Comment by Earlopain** [comment, marked as answer]
```rb
pp  'Some text: %s' %
    [1, 2, 3].map do
      never_called
    end.join("\n")
```

This outputs `"Some text: #<Enumerator:0x214b342f>"` and then complains that string doesn't respond to `join`.
It seems like the block is simply discarded. I'm not sure if that is reasonable for ruby to do and would love some input from others if this may be is a ruby bug.

Created an issue: https://github.com/rubocop/rubocop/issues/13299

**Reply by stdedos** [reply]
What about

I have tried also with

```yaml
Style/BlockDelimiters:
  EnforcedStyle: braces_for_chaining
```

but then I get a different instance of error:

```diff
@@ -92,x +92,y @@ class A < B::C
   end
 
   def parse_values(my_values)
-    my_values.map do |v|
-      begin
+    my_values.map { |v|
+      
         ...
         h
-      rescue => e
+      rescue StandardError => e
         Rails.logger.error("Error parsing report value #{v}: #{e.stacktrace}")
         nil
-      end
-    end.compact
+      
+    }.compact
   end
```

```ruby
org.jruby.exceptions.SyntaxError: (SyntaxError) f.rb:111: syntax error, unexpected `rescue'
      rescue StandardError => e
      ^~~~~~
```
Rubocop:
```ruby
f.rb:95:19: C: [Corrected] Style/BlockDelimiters: Prefer {...} over do...end for multi-line chained blocks.
    my_values.map do |v|
                  ^^
f.rb:96:7: C: [Corrected] Style/RedundantBegin: Redundant begin block detected.
      begin
      ^^^^^
f.rb:101:41: C: [Corrected] Style/IfUnlessModifier: Modifier form of if makes the line too long.
        h.asdfgh_dfgh = vp.dfghjj_fghjk if vp.respond_to?(:cvbnmn_ghjgl) && v.dgsdff_fgje.blank? && vp.ghjklm_yhjkl.present?
                                        ^^
f.rb:109:7: C: [Corrected] Style/RescueStandardError: Avoid rescuing without specifying an error class.
      rescue => e
      ^^^^^^
f.rb:111:7: F: Lint/Syntax: unexpected token kRESCUE
(Using Ruby 2.7 parser; configure using TargetRubyVersion parameter, under AllCops)
      rescue StandardError => e
      ^^^^^^
```

?

Same issue or different?

**Reply by Earlopain** [reply]
Pretty sure its different. Please make an issue if you can. Probably this is a problem with `ensure`/`else` for implicit begin as well.

Note: Providing a diff like that is pretty inconvenient, especially if you don't have the original code. Just the code before correction is good enough.

**Reply by stdedos** [reply]
I was inspired by your issue, and I will try to do the same.

However, for a quick "is this (the same) bug or feature", I thought it illustrated the change well (and the cop(s) involved) 🙏

**Reply by stdedos** [reply]
_Fix incompatible autocorrect between `Style/RedundantBegin` and `Style/BlockDelimiters` with `EnforcedStyle: braces_for_chaining`_

:point_up:  https://github.com/rubocop/rubocop/issues/13302 / https://github.com/rubocop/rubocop/pull/13304
